lib/Conversion/TorchToLinalg/AdaptivePooling.cpp

```diff
@@ -0,0 +1,237 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// Also available under a BSD-style license. See LICENSE.
+//
+//===----------------------------------------------------------------------===//
+
+#include "torch-mlir/Conversion/TorchToLinalg/TorchToLinalg.h"
+
+#include "../PassDetail.h"
+#include "PopulatePatterns.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/ControlFlow/IR/ControlFlowOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/Matchers.h"
+#include "torch-mlir/Conversion/TorchToLinalg/Utils.h"
+#include "torch-mlir/Conversion/Utils/Utils.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchDialect.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchOps.h"
+#include "torch-mlir/Dialect/Torch/Utils/TorchUpstream.h"
+#include "torch-mlir/Dialect/Torch/Utils/Utils.h"
+
+using namespace mlir;
+using namespace mlir::torch;
+using namespace mlir::torch::Torch;
+
+/*
+This file is for lowering adaptive pooling ops, which cannot generally be
+decomposed into typical pooling ops. Given an input tensor of rank (N,C,Hin) and
+an output spatial size Hout, an element of the output tensor at position (n, c,
+h) is computed as follows.
+    1. compute st(h) = (h*Hin)//Hout
+    2. compute en(h) = 1 + ((h+1)*Hin - 1)//Hout
+    3. apply the operation (max or avg) over input[n, c, st(h):en(h)]
+This is problematic for linalg ops for a few reasons:
+    1. The access to the input tensor is not constantly strided
+    2. The size of the window itself is not contant: en(h) - st(h) can vary with
+h! Although it is a bit like using a hammer to paint, our workaround is to use
+tensor.extract to access the elements of the input tensor inside our linalg
+generic op's payload.
+
+Current TODO's:
+    1. roll st(h) and en(h) into affine maps for aesthetics
+    2. gather most of the boilerplate out of this op and make it into an
+adaptive pooling helper function.
+    3. figure out what to do with the conflicting decompositions in
+DecomposeComplexOps.cpp
+    4. Implement more efficient passes for when the kernel-size, input spatial
+dims, and output spatial dims are constant.
+*/
+
+namespace {
+class ConvertAtenAdaptiveAvgPool1dOp
+    : public OpConversionPattern<AtenAdaptiveAvgPool1dOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(AtenAdaptiveAvgPool1dOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Location loc = op->getLoc();
+    const TypeConverter *typeconverter = getTypeConverter();
+
+    // input tensor and output shape
+    Value input = adaptor.getSelf();
+    Value outputShape = adaptor.getOutputSize();
```

Comment:
In MLIR conversions, the `adaptor` gives access to the operands in the dialect _after_ conversion (if it's possible to convert the operand). For example, `op.getSelf()` returns a `!torch.vtensor<...>`, but `adaptor.getSelf()` returns a `tensor<...>`.

This means that for helper functions that work in the torch dialect world, any arguments used should be obtained from `op` instead of `adaptor`. (In this case, `getListConstructElements` works at the torch dialect level, so you should pass `op.getOutputSize()`).

Note: things work here because there is no conversion for `!torch.list`s, so they stay in the torch dialect world, but it is still a good practice to use `op` for the input dialect and `adaptor` for the target dialect.

Reply:
Yeah, that's a good point. I was playing with this to see what the adaptor would give, and forgot to change it back.